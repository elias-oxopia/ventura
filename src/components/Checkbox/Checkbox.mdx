---
name: Checkbox
menu: Components
route: /checkbox
---

import { Playground, Props } from 'docz';
import { Checkbox } from '../../index.ts'
import styles from './docAssets/docs.module.css';

# Checkbox

A checkbox can be used enable an optional setting or feature.

## Best practices

- Checkboxes should be wrapped in a [FormField](/form-field) or should be preceded by another component serving as a label

## Examples

### Basic usage

<Playground>
  <Checkbox name="enabled">Enable me</Checkbox>
</Playground>

### Checked state

<Playground>
  <Checkbox name="enabled-2" defaultValue={true}>Disable me</Checkbox>
</Playground>

### Checkbox with description

<Playground>
  <Checkbox name="enabled-3" description="This feature can be enabled">Enable me</Checkbox>
</Playground>

### Disabled checkbox

<Playground>
  <Checkbox name="enabled-4" isDisabled>You can't enable me</Checkbox>
</Playground>

### Listening for changes

<Playground>
  <Checkbox name="enabled-5" onChange={(event) => console.log(`LOGGER: ${event.target.checked}`)}>Log</Checkbox>
</Playground>

## API

<Props of={Checkbox} />
