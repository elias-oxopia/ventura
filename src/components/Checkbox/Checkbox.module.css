.container {
  display: grid;
  grid-template-areas: 'box title' 'box description';
  grid-template-columns: min-content auto;
  cursor: pointer;
}

.box {
  grid-area: box;
  content: '';
  height: 17px;
  width: 17px;
  border: 2px solid var(--color-neutral-7);
  margin-right: 10px;
  border-radius: 4px;
  cursor: pointer;
}

.isChecked .box {
  background-color: var(--color-neutral-7);
  background-position: center;
  background-repeat: no-repeat;
  background-size: 70%;
  border: none;
}

.isDisabled .box {
  border-color: var(--color-neutral-4);
}

.isChecked.isDisabled .box {
  background-color: var(--color-neutral-4);
}

.input {
  opacity: 0;
  position: absolute;
  cursor: pointer;
}

.title {
  grid-area: title;
  display: block;
  font-weight: 500;
  font-size: 15px;
}

.description {
  grid-area: description;
  display: block;
  color: var(--color-neutral-6);
  margin-top: 8px;
}

.isDisabled .title,
.isDisabled .description {
  color: var(--color-neutral-5);
}
