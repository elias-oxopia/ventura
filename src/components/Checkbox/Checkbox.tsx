import React, { ChangeEventHandler, FocusEventHandler, ReactNode } from 'react';
import classNames from 'classnames';

import cssReset from '../../css-reset.module.css';
import styles from './Checkbox.module.css';

interface Props {
  name: string;
  value?: boolean;
  defaultValue?: boolean;
  onChange?: ChangeEventHandler<HTMLInputElement>;
  onBlur?: FocusEventHandler<HTMLInputElement>;
  className?: string;
  isDisabled?: boolean;
  description?: boolean;
  children: ReactNode;
}

const Input = React.forwardRef<HTMLInputElement, Props>(
  (
    {
      name,
      value,
      defaultValue,
      onChange,
      onBlur,
      isDisabled = false,
      className,
      description,
      children,
    }: Props,
    ref,
  ) => {
    const labelClassNames = classNames(
      cssReset.ventura,
      styles.checkbox,
      {
        [styles.isDisabled]: isDisabled,
      },
      className,
    );

    return (
      <label htmlFor={name} className={labelClassNames}>
        <input
          type="checkbox"
          id={name}
          name={name}
          checked={value}
          defaultChecked={defaultValue}
          onChange={onChange}
          ref={ref}
          data-testid={name && `checkbox-${name}`}
          onBlur={onBlur}
          disabled={isDisabled}
          className={styles.input}
        />
        <div className={styles.text}>
          <span className={styles.title}>{children}</span>
          {description && <span className={styles.description}>{description}</span>}
        </div>
      </label>
    );
  },
);

export default Input;
