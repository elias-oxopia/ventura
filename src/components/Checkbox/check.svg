<svg viewBox="0 0 11 8" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M1 3.82353L3.11641 6.15781C3.52138 6.60447 4.22657 6.59387 4.61793 6.13524L9 1" stroke="white" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"/>
</svg>
