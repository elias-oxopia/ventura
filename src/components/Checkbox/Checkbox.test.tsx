import React, { ComponentProps } from 'react';
import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Checkbox from './Checkbox';

describe('Checkbox', () => {
  const defaultButtonProps: ComponentProps<typeof Checkbox> = {
    name: 'firstname',
  };

  test('default snapshot', () => {
    const component = <Input {...defaultButtonProps} />;
    const { asFragment } = render(component);
    expect(asFragment()).toMatchSnapshot();
  });

  test('value prop', () => {
    const component = <Input {...defaultButtonProps} value="Robin" onChange={() => {}} />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    expect(inputElement.getAttribute('value')).toBe('Robin');
  });

  test('isDisabled prop', async () => {
    const onChangeFn = jest.fn();
    const firstName = 'Bram';
    const component = <Input {...defaultButtonProps} isDisabled onChange={onChangeFn} />;
    const { asFragment, getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    expect(asFragment()).toMatchSnapshot();

    await userEvent.type(inputElement, firstName);
    expect(onChangeFn).toHaveBeenCalledTimes(0);
  });

  test('onChange prop', async () => {
    const onChangeFn = jest.fn();
    const firstName = 'Robin';
    const component = <Input {...defaultButtonProps} onChange={onChangeFn} />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    await userEvent.type(inputElement, firstName);

    expect(onChangeFn).toHaveBeenCalledTimes(firstName.length);
    expect((inputElement as any).value).toBe(firstName);
  });

  test('onBlur prop', async () => {
    const onBlurFn = jest.fn();
    const component = <Input {...defaultButtonProps} onBlur={onBlurFn} />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    fireEvent.blur(inputElement);

    expect(onBlurFn).toHaveBeenCalledTimes(1);
  });

  test('placeholder prop', () => {
    const component = <Input {...defaultButtonProps} placeholder="First name" />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    expect(inputElement.getAttribute('placeholder')).toBe('First name');
  });

  test('type prop', () => {
    const component = <Input {...defaultButtonProps} type="password" />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    expect(inputElement.getAttribute('type')).toBe('password');
  });

  test('errorMessage prop', () => {
    const component = <Input {...defaultButtonProps} isInvalid />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    expect(inputElement.classList).toContain('containsError');
  });

  test('spellCheck prop', () => {
    const component = <Input {...defaultButtonProps} spellCheck />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    expect(inputElement.hasAttribute('spellcheck')).toBe(true);
    expect(inputElement.getAttribute('spellcheck')).not.toBe(false);
  });

  test('autoComplete prop', () => {
    const component = <Input {...defaultButtonProps} autoComplete />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    expect(inputElement.getAttribute('autocomplete')).toBe('on');
  });

  test('maxLength prop', () => {
    const maxLength = 3;
    const component = <Input {...defaultButtonProps} maxLength={3} />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    expect(inputElement.getAttribute('maxlength')).toBe(maxLength.toString());
  });

  test('className prop', () => {
    const className = 'center';
    const component = <Input {...defaultButtonProps} name="classname" className={className} />;
    const { getByTestId } = render(component);
    const containerElement = getByTestId('input-classname');

    const renderedClassNames = containerElement.className.split(' ');
    expect(renderedClassNames).toContain(className);
    // className in prop should be the last in the row
    expect(renderedClassNames.indexOf(className)).toBe(renderedClassNames.length - 1);
  });
});
