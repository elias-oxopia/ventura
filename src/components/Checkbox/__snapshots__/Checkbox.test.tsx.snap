// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox default checked snapshot 1`] = `
<DocumentFragment>
  <label
    class="ventura container isChecked"
    data-testid="checkbox-label-enabled"
    for="enabled"
  >
    <input
      checked=""
      class="input"
      data-testid="checkbox-enabled"
      id="enabled"
      name="enabled"
      type="checkbox"
    />
    <div
      class="box"
      style="background-image: url(undefined);"
    />
    <span
      class="title"
    >
      Enable me
    </span>
  </label>
</DocumentFragment>
`;

exports[`Checkbox default snapshot 1`] = `
<DocumentFragment>
  <label
    class="ventura container"
    data-testid="checkbox-label-enabled"
    for="enabled"
  >
    <input
      class="input"
      data-testid="checkbox-enabled"
      id="enabled"
      name="enabled"
      type="checkbox"
    />
    <div
      class="box"
    />
    <span
      class="title"
    >
      Enable me
    </span>
  </label>
</DocumentFragment>
`;

exports[`Checkbox isDisabled prop 1`] = `
<DocumentFragment>
  <label
    class="ventura container isDisabled"
    data-testid="checkbox-label-enabled"
    for="enabled"
  >
    <input
      class="input"
      data-testid="checkbox-enabled"
      disabled=""
      id="enabled"
      name="enabled"
      type="checkbox"
    />
    <div
      class="box"
    />
    <span
      class="title"
    >
      Enable me
    </span>
  </label>
</DocumentFragment>
`;
